crates/types/src/replication/balanced_spread_selector.rs

```diff
@@ -0,0 +1,979 @@
+                &replication,
+                worker_candidate_filter,
+                &options,
+            )
+            .unwrap();
+
+            assert_that!(nodeset.len(), eq(5));
+            let topology =
+                Topology::from_nodes_configuration(&nodes_config, worker_candidate_filter);
+            assert_that!(
+                topology.check_replication_property(&nodeset, &replication),
+                eq(true)
+            );
+            leaders.insert(nodeset.first().unwrap());
+        }
+        // check how many leaders, we should see diversity
+        assert_that!(leaders.len(), eq(5));
+    }
+
+    #[test]
+    fn select_workers_flat_with_location() {
+        let mut nodes_config = NodesConfiguration::default();
+        for i in 1..=24 {
+            // three regions, 2 zones per region
+            let region = i % 3;
+            let zone = i % 2;
+            // N1 has no location
+            if i == 1 {
+                nodes_config.upsert_node(generate_node(i, WorkerState::Active, Role::Worker, ""));
+            }
+            // N6, 8, 10 are not assigned a zone.
+            if i == 6 || i == 8 || i == 10 {
+                nodes_config.upsert_node(generate_node(
+                    i,
+                    WorkerState::Active,
+                    Role::Worker,
+                    &format!("region-{region}"),
+                ));
+            } else {
+                nodes_config.upsert_node(generate_node(
+                    i,
+                    WorkerState::Active,
+                    Role::Worker,
+                    &format!("region-{region}.zone-{zone}"),
+                ));
+            }
+        }
+
+        let replication =
+            ReplicationProperty::with_scope(LocationScope::Node, 5.try_into().unwrap());
+
+        let mut combined = HashSet::new();
+        // generate for 5 partitions
+        let mut leaders = HashSet::new();
+        for id in 1..=10 {
+            let options = SelectorOptions::new(id);
+            let nodeset = BalancedSpreadSelector::select(
+                &nodes_config,
+                &replication,
+                worker_candidate_filter,
+                &options,
+            )
+            .unwrap();
+
+            assert_that!(nodeset.len(), eq(5));
+            let topology =
+                Topology::from_nodes_configuration(&nodes_config, worker_candidate_filter);
+            assert_that!(
+                topology.check_replication_property(&nodeset, &replication),
+                eq(true)
+            );
+            leaders.insert(nodeset.first().unwrap());
+            combined.extend(nodeset.into_iter());
+        }
+        // check how many leaders, we should see diversity
+        assert_that!(leaders.len(), eq(9));
+        assert_that!(
+            combined,
+            superset_of([
+                PlainNodeId::new(1),
+                PlainNodeId::new(6),
+                PlainNodeId::new(10)
+            ])
```

Comment:
Why are 1, 6 and 10 included in the union of all node sets and not 2, for example? Is this the result of running the algorithm with the given salt?